File: src/app/services/category.service.ts
import { Injectable, Inject } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { APP_CONFIG } from '../../main';

@Injectable({
  providedIn: 'root'
})
export class CategoryService {
  private apiUrl: string;

  constructor(private http: HttpClient, @Inject(APP_CONFIG) private config: any) {
    this.apiUrl = `${this.config.apiUrl}/categories`;
  }

  getAll(): Observable<any[]> {
    return this.http.get<any[]>(`${this.apiUrl}/all`);
  }

  getById(id: number): Observable<any> {
    return this.http.get<any>(`${this.apiUrl}/${id}`);
  }

  create(category: any): Observable<any> {
    return this.http.post<any>(`${this.apiUrl}/create`, category);
  }

  update(id: number, category: any): Observable<any> {
    return this.http.put<any>(`${this.apiUrl}/update/${id}`, category);
  }

  delete(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/delete/${id}`);
  }
}

File: lib/services/category_service.dart
import '../models/category.dart';
import '../utils/constants.dart';
import 'api_service.dart';

class CategoryService {
  final ApiService _apiService = ApiService();
  final String _baseUrl = '${AppConstants.apiUrl}/categories';

  Future<List<Category>> getAll() async {
    final resp = await _apiService.get('$_baseUrl/all');
    if (resp is List) {
      return resp.map((e) => Category.fromJson(Map<String, dynamic>.from(e as Map))).toList();
    }
    return [];
  }

  Future<Category> create(Map<String, dynamic> payload) async {
    final resp = await _apiService.post('$_baseUrl/create', payload);
    return Category.fromJson(Map<String, dynamic>.from(resp as Map));
  }

  Future<Category> update(int id, Map<String, dynamic> payload) async {
    final resp = await _apiService.put('$_baseUrl/update/$id', payload);
    return Category.fromJson(Map<String, dynamic>.from(resp as Map));
  }

  Future<void> delete(int id) async {
    await _apiService.delete('$_baseUrl/delete/$id');
  }
}
